analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs

```diff
@@ -552,5 +552,43 @@ public void BinaryNotEqualsNull_SetsBoolConstraint_Unknown_ComparedToNull_VB(str
                 .And.ContainSingle(state => state.SymbolsWith(BoolConstraint.True).Any(x => x.Name == "Value") && state.SymbolsWith(ObjectConstraint.NotNull).Any(x => x.Name == "Arg"))
                 .And.ContainSingle(state => state.SymbolsWith(BoolConstraint.False).Any(x => x.Name == "Value") && state.SymbolsWith(ObjectConstraint.Null).Any(x => x.Name == "Arg"));
         }
+
+        [DataRow("arg >= null")]
+        [DataRow("arg > null")]
+        [DataRow("arg < null")]
+        [DataRow("arg <= null")]
+        [DataRow("null >= arg")]
+        [DataRow("null > arg")]
+        [DataRow("null < arg")]
+        [DataRow("null <= arg")]
+        [DataRow("arg > (int?)null")]
+        [DataRow("arg > new Nullable<int>()")]
+        [DataRow("arg > (null as int?)")]
+        [DataRow("arg > nullValue")]
+        [DataRow("nullValue > arg")]
+        [DataRow("nullValue > 42")]
+        [DataRow("nullValue > notNullValue")]
+        [DataRow("notNullValue > nullValue")]
+        [TestMethod]
+        public void Binary_NullableRelationalNull_SetsBoolConstraint_CS(string expression)
```

Comment:
This test does not directly test the changes of this PR.

Reply:
The changes in this PR alter the behavior of `BoolConstraintFromOperation`, which is responsible to inform the branching processor on whether the result of the branching condition is known with certainty, which means no branching of the `ProgramState` is needed, or not.
```cs
protected override ProgramState[] Process(SymbolicContext context, T operation)
{
    var state = PreProcess(context.State, operation);
    if (BoolConstraintFromOperation(state, operation) is { } constraint)
    {
        return state.SetOperationConstraint(context.Operation, constraint).ToArray();    // We already know the answer from existing constraints
    }
    else
    {
        var beforeLearningState = state;
        var positive = LearnBranchingConstraint(state, operation, false);
        var negative = LearnBranchingConstraint(state, operation, true);
        return positive == beforeLearningState && negative == beforeLearningState
            ? beforeLearningState.ToArray()   // We can't learn anything, just move on
            : new[]
            {
                positive.SetOperationConstraint(context.Operation, BoolConstraint.True),
                negative.SetOperationConstraint(context.Operation, BoolConstraint.False)
            };
    }
}
```

The `BoolConstraint` is on the branching condition, and the easiest way I can think of is to ensure that we enter the "then" branch of the if above, instead of entering the "else" branch, which means no `ProgramState` branching.

So, by checking that I never go through `If_Unreachable`, I am ensuring that the `BoolConstraint` is set on the branching c condition, when a relational operator is used against `null`, which is what this PR is about.

Do you see more direct ways of testing this?

Reply:
I have changed the title of the PR into `SE: Don't branch on relational operator against null`, which is more precise that before. Does that help?

Reply:
[DRAFT]
Learning constraints from binary operations is done in `LearnBranchingRelationalConstraint`, which is called when the evaluation of the branching condition is not conclusive (i.e. we don't know for certain whether the condition is true or false, and we need to branch the state).
This PR only affects `BoolConstraintFromOperation`

Reply:
As discussed offline, both test strategies are testing the constraint indirectly, from two different sides.

Added a test as proposed in https://github.com/SonarSource/sonar-dotnet/pull/7054#discussion_r1165302586, sharing the same test data.